index.d.ts

```diff
@@ -0,0 +1,36 @@
+import { Action } from 'redux';
+
+export type Enhancement = object;
+
+export type EnhancerId = string | symbol;
+
+type ActionEnhancerById = {
+  readonly id: EnhancerId;
+};
+
+type ActionEnhancerByType = {
+  actionTypes: string[];
+};
+
+export type ActionEnhancer<S, E extends Enhancement, V = any> = (
+  | ActionEnhancerById
+  | ActionEnhancerByType) & {
+  mapState: (state: S, value: V) => E;
+};
+
+export type UnenhancedAction<T> = Action<T> & {
+  [index: string]: any;
+};
+
+export type EnhancedAction<
+  A extends Action,
+  E1 extends Enhancement,
+  E2 extends Enhancement = E1,
+  E3 extends Enhancement = E1,
+  E4 extends Enhancement = E1,
+  E5 extends Enhancement = E1,
+  E6 extends Enhancement = E1,
```

Comment:
should these default to undefined or something instead? or is it correct to  just repeat E1 over and over? 

What happens if there's a conflict between 2 enhancers? Let's say both set the property foo (obviously that won't work without ts either, i'm just curious if the type system will complain)

Reply:
To the former; it's fine to repeat E1 over and over. Using `undefined` or something would mess up the union type. With `E1` repeated TypeScript collapses the duplicates.

Here's an example:
![image](https://user-images.githubusercontent.com/52676728/80509911-a0af2800-893f-11ea-990a-32aec6ac6f73.png)

To the latter..

1. If two enhancers specify the same field name with the same type, TypeScript doesn't say anything.
2. If two enhancers specify the same field name with different types, TypeScript ends up marking that field as a `never` (which is a bit surprising) and if you try to access the value the compiler throws errors.